clang/test/CodeGen/X86/x86-builtins-intrinsic.cl

```diff
@@ -0,0 +1,73 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
+// RUN:  %clang_cc1 -no-opaque-pointers %s -triple x86_64-pc-linux  \
+// RUN:   -emit-llvm  -O0  -x cl -cl-std=CL2.0 -ffake-address-space-map \
+// RUN:   -target-cpu skx -o - | FileCheck %s
+
+typedef int __v16si __attribute__((__vector_size__(64)));
+
+// CHECK-LABEL: @load(
+// CHECK-NEXT:  entry:
+// CHECK-NEXT:    [[__U_ADDR:%.*]] = alloca i16, align 2
+// CHECK-NEXT:    [[__P_ADDR:%.*]] = alloca i8 addrspace(4)*, align 8
+// CHECK-NEXT:    [[DD:%.*]] = alloca <16 x i32>, align 64
+// CHECK-NEXT:    store i16 [[__U:%.*]], i16* [[__U_ADDR]], align 2
+// CHECK-NEXT:    store i8 addrspace(4)* [[__P:%.*]], i8 addrspace(4)** [[__P_ADDR]], align 8
+// CHECK-NEXT:    store <16 x i32> zeroinitializer, <16 x i32>* [[DD]], align 64
+// CHECK-NEXT:    [[TMP0:%.*]] = load i8 addrspace(4)*, i8 addrspace(4)** [[__P_ADDR]], align 8
+// CHECK-NEXT:    [[TMP1:%.*]] = bitcast i8 addrspace(4)* [[TMP0]] to i32 addrspace(4)*
+// CHECK-NEXT:    [[TMP2:%.*]] = load <16 x i32>, <16 x i32>* [[DD]], align 64
+// CHECK-NEXT:    [[TMP3:%.*]] = load i16, i16* [[__U_ADDR]], align 2
+// CHECK-NEXT:    [[TMP4:%.*]] = addrspacecast i32 addrspace(4)* [[TMP1]] to <16 x i32>*
```

Comment:
Is address space cast valid if the pointers being casted have not only different address spaces but also different pointee types?
I can't double check with modern LLVM langref since it only covers opaque pointers now.

Reply:
I updated the test case.  please double check.

Reply:
> Is address space cast valid if the pointers being casted have not only different address spaces but also different pointee types?

Originally this is a 
addrspace cast from <16 x i32>* to <16 x i32> addrspace(4) *
followed by a
bitcast from i32 addrspace(4)* to <16 x i32> addrspace(4) *

Now the new code uses CreatePointerBitCastOrAddrSpaceCast. This is probably fine for opaque pointer since type is gone in this line.

Reply:
I suggest until we use typed pointers in our project, still check them it LIT testing.

Reply:
> Now the new code uses CreatePointerBitCastOrAddrSpaceCast. This might be fine for opaque pointer since type is gone in this line.

How about typed pointers? Is it ok for them?

Reply:
> > Now the new code uses CreatePointerBitCastOrAddrSpaceCast. This might be fine for opaque pointer since type is gone in this line.
> 
> How about typed pointers? Is it ok for them?

it should be ok. Following are examples from addrspacecast section in xmain 20210201 LangRef.rst:
      %Y = addrspacecast i32 addrspace(1)* %y to i64 addrspace(2)*    ; yields i64 addrspace(2)*:%y
      %Z = addrspacecast <4 x i32*> %z to <4 x float addrspace(3)*>   ; yields <4 x float addrspace(3)*>:%z